src/mlpack/tests/ann_layer_test.cpp

```diff
@@ -3810,4 +3810,205 @@ BOOST_AUTO_TEST_CASE(TransposedConvolutionalLayerOptionalParameterTest)
+  arma::mat gy = arma::ones(embedDim * tLen, batchSize);
+  arma::mat g;
+  module.Backward(input, gy, g);
+  arma::mat expGrad = arma::mat("0.010093 0.051616;\
+                                 0.011682 0.021405;\
+                                -0.009071 -0.014949;\
+                                 0.028626 -0.005551;\
+                                 0.014471 0.051683;\
+                                 0.011908 0.021948;\
+                                -0.012031 -0.015816;\
+                                 0.028747 -0.005957;\
+                                 0.011069 0.051405;\
+                                 0.012144 0.023553;\
+                                -0.009494 -0.016044;\
+                                 0.028043 -0.006889;\
+                                 0.012005 0.052027;\
+                                 0.010873 0.022022;\
+                                -0.010914 -0.015380;\
+                                 0.030171 -0.005879");
+
+  for (size_t i = 0; i < batchSize; ++i)
+  {
+    for (size_t j = 0; j < expGrad.n_rows; ++j)
+    {
+      BOOST_REQUIRE_CLOSE_FRACTION(g(j, i), expGrad(j, i), 1e-04);
+    }
+  }
+
+  BOOST_REQUIRE_EQUAL(g.n_rows, input.n_rows);
+  BOOST_REQUIRE_EQUAL(g.n_cols, input.n_cols);
+
+  //! Scaled Dot Product Attention test with attention and key padding mask.
+  double INF = 1e+09;
+  arma::mat attnMask = arma::zeros(sLen, tLen);
+  for (size_t i = 0; i < sLen; ++i)
+  {
+    for (size_t j = 0; j < tLen; ++j)
+    {
+      if (i > j)
+        attnMask(i, j) = -INF;
+    }
+  }
+
+  arma::mat keyPaddingMask = arma::zeros(sLen, 1);
+  keyPaddingMask(sLen - 2) = -INF;
+
+  module.AttentionMask() = attnMask;
+  module.KeyPaddingMask() = keyPaddingMask;
+
+  //! Test Forward function with mask.
+  module.Forward(input, output);
+  expOutput = arma::mat("0.777000 0.362100;\
+                         0.583900 0.736400;\
+                         0.950300 0.519400;\
+                         0.438100 0.423000;\
+                         0.549956 0.178589;\
+                         0.558185 0.831599;\
+                         0.603817 0.295066;\
+                         0.242064 0.729971;\
+                         0.559399 0.185750;\
+                         0.559254 0.827884;\
+                         0.618227 0.303820;\
+                         0.250217 0.717994;\
+                         0.393847 0.306477;\
+                         0.434156 0.696050;\
+                         0.723690 0.488940;\
+                         0.360290 0.774306");
+
+  for (size_t i = 0; i < output.n_cols; ++i)
+  {
+    for (size_t j = 0; j < output.n_rows; ++j)
+      BOOST_REQUIRE_CLOSE_FRACTION(output(j, i), expOutput(j, i), 1e-04);
+  }
+
+  BOOST_REQUIRE_EQUAL(output.n_rows, embedDim * tLen);
+  BOOST_REQUIRE_EQUAL(output.n_cols, input.n_cols);
+
+  //! Test Backward function with mask.
+  module.Backward(input, gy, g);
+  expGrad = arma::mat("0.00000000 0.00000000;\
```

Comment:
By the way, do you have a notebook for this? I'd like to play with the values a bit.

Reply:
Yup. [Here](https://colab.research.google.com/drive/1QKJe9WLxAMn9LWWMt2Pyp2p0_DsHRThd) it is.